// mdb-cli/config/index.js
'use strict';
Object.defineProperty(exports, "__esModule", { value: true });
var dotenv_1 = require("dotenv");
var os_1 = require("os");
var path_1 = require("path");
dotenv_1.config({ path: path_1.join(__dirname, '..', '.env') });
var env = process.env.NODE_ENV || 'prd';
var gitlabUrl = process.env.GITLAB_URL || 'https://git.mdbgo.com';
var projectsDomain = process.env.PROJECTS_DOMAIN || 'mdbgo.io';
var technologies = process.env.BACKEND_TECHNOLOGIES || 'node8,node10,node12,node14,node16,node17,php7.2,php7.3,php7.4,php-laravel,python3.10';
var databases = process.env.DATABASES || 'mysql8,mongodb';
var mdbgoPipelinePublicBranch = process.env.MDBGO_PIPELINE_PUBLIC_BRANCH || 'mdbgo/public';
var memberRoles = process.env.MEMBER_ROLES || 'owner,developer,reporter';
var host = process.env.HOST || 'apps.mdbootstrap.com';
var port = process.env.PORT;
var apiPath = typeof process.env.API_PATH === "undefined" ? '/api' : process.env.API_PATH;
process.env['NODE_NO_WARNINGS'] = '1';
exports.default = {
    env: env,
    gitlabUrl: gitlabUrl,
    mdbgoPipelinePublicBranch: mdbgoPipelinePublicBranch,
    projectsDomain: projectsDomain,
    databases: databases.split(','),
    memberRoles: memberRoles.split(','),
    tokenFile: '.auth',
    tokenDir: path_1.join(os_1.homedir(), '.mdbcli'),
    msgPath: path_1.join(os_1.homedir(), '.mdbcli', '.msg'),
    port: port,
    host: host,
    apiPath: apiPath,
    apiUrl: env === 'dev' ? process.env.API_URL : "https://" + host + apiPath,
    auth: {
        social: {
            google: {
                url: process.env.AUTH_SOCIAL_GOOGLE_URL
            },
            facebook: {
                url: process.env.AUTH_SOCIAL_FACEBOOK_URL
            },
            twitter: {
                url: process.env.AUTH_SOCIAL_TWITTER_URL
            }
        }
    },
    backend: {
        technologies: technologies.split(','),
        requiredFiles: [
            { tech: 'node', name: 'package.json' },
            { tech: 'laravel', name: 'composer.json' },
            { tech: 'python', name: 'requirements.txt' }
        ]
    },
    projectTypes: ['frontend', 'backend', 'wordpress']
};
